runtime/src/bank.rs

```diff
@@ -2715,6 +2726,122 @@ impl Bank {
             vote_with_stake_delegations_map,
             invalid_vote_keys,
             invalid_stake_keys,
+            vote_accounts_cache_miss_count: 0,
+        }
+    }
+
+    fn load_vote_and_stake_accounts<F>(
+        &self,
+        thread_pool: &ThreadPool,
+        reward_calc_tracer: Option<F>,
+    ) -> LoadVoteAndStakeAccountsResult
+    where
+        F: Fn(&RewardCalculationEvent) + Send + Sync,
+    {
+        let stakes = self.stakes_cache.stakes();
+        let stake_delegations: Vec<_> = stakes.stake_delegations().iter().collect();
+        // Obtain all unique voter pubkeys from stake delegations.
+        fn merge(mut acc: HashSet<Pubkey>, other: HashSet<Pubkey>) -> HashSet<Pubkey> {
+            if acc.len() < other.len() {
+                return merge(other, acc);
+            }
+            acc.extend(other);
+            acc
+        }
+        let voter_pubkeys = thread_pool.install(|| {
+            stake_delegations
+                .par_iter()
+                .fold(
+                    HashSet::default,
+                    |mut voter_pubkeys, (_stake_pubkey, stake_account)| {
+                        let delegation = stake_account.delegation().unwrap();
+                        voter_pubkeys.insert(delegation.voter_pubkey);
+                        voter_pubkeys
+                    },
+                )
+                .reduce(HashSet::default, merge)
+        });
+        // Obtain vote-accounts for unique voter pubkeys.
+        let cached_vote_accounts = stakes.vote_accounts();
+        let solana_vote_program: Pubkey = solana_vote_program::id();
+        let vote_accounts_cache_miss_count = AtomicUsize::default();
+        let get_vote_account = |vote_pubkey: &Pubkey| -> Option<VoteAccount> {
+            if let Some((_stake, vote_account)) = cached_vote_accounts.get(vote_pubkey) {
+                return Some(vote_account.clone());
+            }
+            // If accounts-db contains a valid vote account, then it should
+            // already have been cached in cached_vote_accounts; so the code
+            // below is only for sanity check, and can be removed once
+            // vote_accounts_cache_miss_count is shown to be always zero.
+            let account = self.get_account_with_fixed_root(vote_pubkey)?;
+            if account.owner() == &solana_vote_program
+                && VoteState::deserialize(account.data()).is_ok()
```

Comment:
should we reuse this work to construct the `VoteAccount`?

Reply:
The code already has comments here that this is only for sanity check and this path should never return a valid vote-account. My intention is to remove this piece all together once it is confirmed `vote_accounts_cache_miss_count` is always zero.

Reply:
@behzadnouri I missed this earlier but I think that the code here will produce false positives for the `vote_accounts_cache_miss_count` counter. We only add vote accounts to the cache if they are initialized so it's possible that an uninitialized vote account has a stake delegation and deserializes correctly. If that happens, this counter would indicate that we had a cache miss but the cache was actually correct. This all stems from the fact that the vote state doesn't have an intuitive "uninitialized" state. Best we can do is check if the authorized voter is non-zero:

https://github.com/solana-labs/solana/blob/7cd7173b71ba80a8698743c6289e6489844376c8/programs/vote/src/vote_state/vote_state_versions.rs#L54-L62

Reply:
So then shouldn't this code also check if the vote-account is initialized?
https://github.com/solana-labs/solana/blob/30d2b112e/runtime/src/bank.rs#L2848-L2888

Seems to me either we should skip uninitialized vote accounts there as well, or store them in the cache as well.

Reply:
Yes, exactly. I brought this up because I think this code needs to check for initialization but it isn't doing that right now.